generator.py

```diff
@@ -0,0 +1,233 @@
+        else:
+            if is_sys_crate:
+                args.extend(["-m", "sys"])
+
+            async def regenerate_crate(path, args):
+                return "Regenerating `{}`...\n".format(path) + await spawn_gir(
+                    conf.gir_path, args
+                )
+
+            processes.append(regenerate_crate(path, args))
+
+    else:
+        raise Exception("`{}` is not a valid Gir*.toml file".format(path))
+
+    return processes
+
+
+def valid_path(path):
+    path = Path(path)
+    if not path.exists():
+        raise argparse.ArgumentTypeError("`{}` no such file or directory".format(path))
+    return path
+
+
+def directory_path(path):
+    path = Path(path)
+    if not path.is_dir():
+        raise argparse.ArgumentTypeError("`{}` directory not found".format(path))
+    return path
+
+
+def file_path(path):
+    path = Path(path)
+    if not path.is_file():
+        raise argparse.ArgumentTypeError("`{}` file not found".format(path))
+    return path
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(
+        description="Helper to regenerate gtk-rs crates using gir.",
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
+    )
+
+    parser.add_argument(
+        "path",
+        nargs="*",
+        default=[Path(".")],
+        type=valid_path,
+        help="Paths in which to look for Gir.toml files",
+    )
+    parser.add_argument(
+        "--gir-files-directories",
+        nargs="+",  # If the option is used, we expect at least one folder!
+        dest="gir_files_paths",
+        default=[],
+        type=directory_path,
+        help="Path of the gir-files folder",
+    )
+    parser.add_argument(
+        "--gir-path",
+        default=DEFAULT_GIR_PATH,
+        type=file_path,
+        help="Path of the gir executable to run",
+    )
+    parser.add_argument(
+        "--yes",
+        action="store_true",
+        help=" Always answer `yes` to any question asked by the script",
+    )
+    parser.add_argument(
+        "--no-fmt",
+        action="store_true",
+        help="If set, this script will not run `cargo fmt`",
+    )
+    parser.add_argument(
+        "--embed-docs",
+        action="store_true",
+        help="Build documentation with `gir -m doc`, and embed it with `rustdoc-stripper -g`",
+    )
+
+    return parser.parse_args()
+
+
+async def main():
+    conf = parse_args()
+
+    if not conf.gir_files_paths:
+        update_submodule(DEFAULT_GIR_FILES_DIRECTORY, conf)
+
+    if conf.gir_path == DEFAULT_GIR_PATH:
+        update_submodule(DEFAULT_GIR_DIRECTORY, conf)
+        build_gir()
+
+    print("=> Regenerating crates...")
+    for path in conf.path:
+        print("=> Looking in path `{}`".format(path))
+        # Collect and print the results as soon as they trickle in, one process at a time:
+        for coro in asyncio.as_completed(regen_crates(path, conf)):
+            print(await coro)
+
+    if not conf.no_fmt and not run_command(["cargo", "fmt"]):
```

Comment:
```suggestion
    if not conf.no_fmt or not run_command(["cargo", "fmt"]):
```

Reply:
I don't think we want to `return 1` when `no_fmt` is `False`? This chains `run_command` if `no_fmt` is false, should we apply some De Morgan instead?

```suggestion
    if not (conf.no_fmt or run_command(["cargo", "fmt"])):
```

Reply:
Perhaps this is clearer to read when separated in two if statements:
```suggestion
    if not conf.no_fmt:
        if not run_command(["cargo", "fmt"]):
```

Reply:
Oh indeed!

Reply:
We can also perform the inversion for `no_fmt` inside `argparse` if you want, with `default=True` and `store_false`?